dex/src/main/scala/com/wavesplatform/dex/SpendableBalancesActor.scala

```diff
@@ -0,0 +1,89 @@
+import com.wavesplatform.dex.domain.account.Address
+import com.wavesplatform.dex.domain.asset.Asset
+import com.wavesplatform.dex.fp.MapImplicits.cleaningGroup
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class SpendableBalancesActor(spendableBalances: (Address, Set[Asset]) => Future[Map[Asset, Long]],
+                             allAssetsSpendableBalances: Address => Future[Map[Asset, Long]],
+                             addressDirectory: ActorRef)(implicit ec: ExecutionContext)
+    extends Actor {
+
+  var fullState: Map[Address, Map[Asset, Long]]  = Map.empty
+  var addressDff: Map[Address, Map[Asset, Long]] = Map.empty
+
+  def receive: Receive = {
+
+    case SpendableBalancesActor.Query.GetState(address, assets) =>
+      val maybeAddressState            = fullState.get(address).orElse(addressDff get address)
+      val assetsMaybeBalances          = assets.map(asset => asset -> maybeAddressState.flatMap(_ get asset)).toMap
+      val (knownAssets, unknownAssets) = assetsMaybeBalances.partition { case (_, balance) => balance.isDefined }
+      val knownPreparedState           = knownAssets.collect { case (a, Some(b)) => a -> b }
+
+      if (unknownAssets.isEmpty) sender ! SpendableBalancesActor.Reply.GetState(knownPreparedState)
+      else {
+        spendableBalances(address, unknownAssets.keySet).map { stateFromNode =>
+          if (fullState contains address) fullState = fullState.updated(address, fullState.getOrElse(address, Map.empty) ++ stateFromNode)
+          else addressDff = addressDff.updated(address, addressDff.getOrElse(address, Map.empty) ++ stateFromNode)
+          SpendableBalancesActor.Reply.GetState(stateFromNode ++ knownPreparedState)
+        } pipeTo sender
+      }
+
+    case SpendableBalancesActor.Query.GetSnapshot(address) =>
+      fullState.get(address) match {
+        case Some(state) => sender ! SpendableBalancesActor.Reply.GetSnapshot(state)
+        case None        => allAssetsSpendableBalances(address).map(SpendableBalancesActor.Command.SetState(address, _)).pipeTo(self)(sender)
+      }
+
+    case SpendableBalancesActor.Command.SetState(address, state) =>
+      val addressState = state ++ addressDff.getOrElse(address, Map.empty)
+      fullState += address -> addressState
+      addressDff -= address
+      sender ! SpendableBalancesActor.Reply.GetSnapshot(addressState)
+
+    case SpendableBalancesActor.Command.UpdateStates(changes) =>
+      changes.foreach {
+        case (address, stateUpdate) =>
+          fullState.get(address) match {
+            case Some(addressFullState) =>
+              val cleanStateUpdate = stateUpdate.filter { case (a, updatedBalance) => addressFullState.get(a).fold(true)(updatedBalance != _) }
+              addressDirectory ! AddressDirectory.Envelope(address, AddressActor.Command.CancelNotEnoughCoinsOrders(cleanStateUpdate))
+              fullState = fullState.updated(address, addressFullState ++ stateUpdate)
```

Comment:
```scala
addressFullState ++ cleanStateUpdate)
```
? Because `cleanStateUpdate` contains:
* Assets those haven't came before;
* Assets with different amounts.

Reply:
Sensible, done